File: data_worker/src/iface/communication/commsec_channel_manager.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from uuid import UUID

from msgspec import Struct

from src.iface.communication.commsec_contract import ChannelEncryptResult, SecureChannelEnsureRequest
from src.models.commsec.commsec import (
    CipherSuite,
    ECDHEHandshakeRecord,
    KeyExchangeAlgorithm,
    SecureChannelBinding,
    SecureChannelSession,
    ServiceKeyOwner,
    SignatureAlgorithm,
)


class ECDHEHandshakeInitRequest(Struct, kw_only=True):
    """ECDHE 初始化请求契约。"""

    initiator: ServiceKeyOwner
    responder: ServiceKeyOwner

    initiator_key_id: str

    supported_key_exchanges: list[KeyExchangeAlgorithm]
    supported_signatures: list[SignatureAlgorithm]
    supported_cipher_suites: list[CipherSuite]

    binding: SecureChannelBinding
    ttl_sec: int


class ECDHEHandshakeInitResult(Struct, kw_only=True):
    """ECDHE 初始化结果契约。"""

    handshake: ECDHEHandshakeRecord

    selected_key_exchange: KeyExchangeAlgorithm
    selected_signature: SignatureAlgorithm
    selected_cipher_suite: CipherSuite


class ECDHEHandshakeCompleteRequest(Struct, kw_only=True):
    """ECDHE 完成请求契约。"""

    handshake_id: UUID

    responder_ephemeral_public_key: str
    responder_signature: str
    responder_nonce: str


class ECDHEHandshakeCompleteResult(Struct, kw_only=True):
    """ECDHE 完成结果契约。"""

    handshake: ECDHEHandshakeRecord
    channel: SecureChannelSession | None = None


class SecureChannelUpsertRequest(Struct, kw_only=True):
    """安全通道写入请求契约。"""

    handshake_id: UUID
    binding: SecureChannelBinding

    source: ServiceKeyOwner
    target: ServiceKeyOwner

    local_key_id: str
    peer_key_id: str

    cipher_suite: CipherSuite
    derived_key_ref: str
    ttl_sec: int


class SecureChannelQuery(Struct, kw_only=True):
    """安全通道查询请求契约。"""

    channel_id: UUID
    binding: SecureChannelBinding

    source_entity_id: str
    target_entity_id: str


class SecureChannelRevokeRequest(Struct, kw_only=True):
    """安全通道撤销请求契约。"""

    channel_id: UUID

    binding: SecureChannelBinding
    reason: str
    revoked_by: str


class ChannelEncryptRequest(Struct, kw_only=True):
    """通过安全通道进行载荷加密的请求契约。"""

    channel_id: UUID
    plaintext: str
    additional_data: dict[str, str] | None = None


class ChannelDecryptRequest(Struct, kw_only=True):
    """通过安全通道进行载荷解密的请求契约。"""

    channel_id: UUID
    ciphertext: str
    sequence: int
    additional_data: dict[str, str] | None = None


@dataclass(slots=True, kw_only=True)
class ChannelDecryptResult:
    plaintext: str
    updated_sequence: int


class ICommsecChannelManager(ABC):
    """安全通道生命周期与载荷操作。

    下游接口调用：
      - common.IKeyManager.get_private_key_ref / lookup_public_key
      - commsec 服务实现使用的密码学原语
    """

    @abstractmethod
    async def ensure_channel(self, req: SecureChannelEnsureRequest) -> SecureChannelSession:
        raise NotImplementedError

    @abstractmethod
    async def init_handshake(self, req: ECDHEHandshakeInitRequest) -> ECDHEHandshakeInitResult:
        raise NotImplementedError

    @abstractmethod
    async def complete_handshake(
        self,
        req: ECDHEHandshakeCompleteRequest,
    ) -> ECDHEHandshakeCompleteResult:
        raise NotImplementedError

    @abstractmethod
    async def upsert_channel(self, req: SecureChannelUpsertRequest) -> SecureChannelSession:
        raise NotImplementedError

    @abstractmethod
    async def get_channel(self, query: SecureChannelQuery) -> SecureChannelSession | None:
        raise NotImplementedError

    @abstractmethod
    async def revoke_channel(self, req: SecureChannelRevokeRequest) -> None:
        raise NotImplementedError

    @abstractmethod
    async def encrypt_payload(self, req: ChannelEncryptRequest) -> ChannelEncryptResult:
        raise NotImplementedError

    @abstractmethod
    async def decrypt_payload(self, req: ChannelDecryptRequest) -> ChannelDecryptResult:
        raise NotImplementedError

File: certification_server/src/iface/communication/commsec_channel_manager.go
package communication

import (
	"context"

	commsecmodel "certification_server/src/models/commsec"

	"github.com/google/uuid"
)

// SecureChannelEnsureRequest 是认证侧的通道确保请求契约。
type SecureChannelEnsureRequest struct {
	Query *SecureChannelQuery

	HandshakeInit *ECDHEHandshakeInitRequest

	RequireActive    bool
	ForceReHandshake bool
}

// ECDHEHandshakeInitRequest 表示通信层的 ECDHE 握手初始化请求。
type ECDHEHandshakeInitRequest struct {
	ChannelClass commsecmodel.ChannelClass
	SecurityMode commsecmodel.ChannelSecurityMode

	Initiator commsecmodel.ServiceKeyOwner
	Responder commsecmodel.ServiceKeyOwner

	InitiatorKeyID string

	SupportedKeyExchanges []commsecmodel.KeyExchangeAlgorithm
	SupportedSignatures   []commsecmodel.SignatureAlgorithm
	SupportedCipherSuites []commsecmodel.CipherSuite

	Binding commsecmodel.SecureChannelBinding
	TTLSec  int64
}

// ECDHEHandshakeInitResult 表示通信层的 ECDHE 握手初始化结果。
type ECDHEHandshakeInitResult struct {
	Handshake commsecmodel.ECDHEHandshakeRecord

	SelectedKeyExchange commsecmodel.KeyExchangeAlgorithm
	SelectedSignature   commsecmodel.SignatureAlgorithm
	SelectedCipherSuite commsecmodel.CipherSuite
}

// ECDHEHandshakeCompleteRequest 表示通信层的 ECDHE 握手完成请求。
type ECDHEHandshakeCompleteRequest struct {
	HandshakeID uuid.UUID

	ResponderEphemeralPublicKey string
	ResponderSignature          string
	ResponderNonce              string
}

// ECDHEHandshakeCompleteResult 是通信层的 ECDHE 握手完成结果。
type ECDHEHandshakeCompleteResult struct {
	Handshake commsecmodel.ECDHEHandshakeRecord
	Channel   *commsecmodel.SecureChannelSession
}

// SecureChannelUpsertRequest 是认证侧的通道创建/更新请求契约。
type SecureChannelUpsertRequest struct {
	HandshakeID  uuid.UUID
	ChannelClass commsecmodel.ChannelClass
	SecurityMode commsecmodel.ChannelSecurityMode
	Binding      commsecmodel.SecureChannelBinding

	Source commsecmodel.ServiceKeyOwner
	Target commsecmodel.ServiceKeyOwner

	LocalKeyID string
	PeerKeyID  string

	CipherSuite   commsecmodel.CipherSuite
	DerivedKeyRef string
	TTLSec        int64
}

// SecureChannelQuery 是通信层的通道查询契约。
type SecureChannelQuery struct {
	ChannelID    uuid.UUID
	Binding      commsecmodel.SecureChannelBinding
	ChannelClass commsecmodel.ChannelClass

	SourceEntityID string
	TargetEntityID string
}

// SecureChannelRevokeRequest 是认证侧的通道撤销请求契约。
type SecureChannelRevokeRequest struct {
	ChannelID uuid.UUID

	Binding   commsecmodel.SecureChannelBinding
	Reason    string
	RevokedBy string
}

// ChannelEncryptRequest 是通信层的载荷加密请求契约。
type ChannelEncryptRequest struct {
	ChannelID      uuid.UUID
	PlainText      string
	AdditionalData map[string]string
}

// ChannelEncryptResult 表示通信层的载荷加密结果。
type ChannelEncryptResult struct {
	CipherText string
	Meta       commsecmodel.EncryptedMessageMeta
}

// ChannelDecryptRequest 是通信层的载荷解密请求契约。
type ChannelDecryptRequest struct {
	ChannelID      uuid.UUID
	CipherText     string
	Sequence       uint64
	AdditionalData map[string]string
}

// ICommsecChannelManager 定义安全通道生命周期与载荷操作。
//
// 下游接口调用：
//   - common.IKeyManager.GetPrivateKeyRef / LookupPublicKey
//   - commsec 服务实现使用的密码学原语
type ICommsecChannelManager interface {
	EnsureChannel(ctx context.Context, req *SecureChannelEnsureRequest) (*commsecmodel.SecureChannelSession, error)

	InitHandshake(ctx context.Context, req *ECDHEHandshakeInitRequest) (*ECDHEHandshakeInitResult, error)
	CompleteHandshake(ctx context.Context, req *ECDHEHandshakeCompleteRequest) (*ECDHEHandshakeCompleteResult, error)

	UpsertChannel(ctx context.Context, req *SecureChannelUpsertRequest) (*commsecmodel.SecureChannelSession, error)
	GetChannel(ctx context.Context, query *SecureChannelQuery) (*commsecmodel.SecureChannelSession, error)
	RevokeChannel(ctx context.Context, req *SecureChannelRevokeRequest) error

	EncryptPayload(ctx context.Context, req *ChannelEncryptRequest) (*ChannelEncryptResult, error)
	DecryptPayload(ctx context.Context, req *ChannelDecryptRequest) (string, error)
}
